File: Java/91. Decode Ways.java
class Solution {
    public int numDecodings(String s) {
        int[] dp = new int[s.length() + 1];
        Arrays.fill(dp, 0);
        dp[0] = 1;
        for (int i = 1; i < s.length() + 1; i++) {
            if (s.charAt(i - 1) >= '1' && s.charAt(i - 1) <= '9')
                dp[i] += dp[i - 1];
            if (i != 1) {
                if (s.charAt(i - 1) >= '0' && s.charAt(i - 1) <= '9' && s.charAt(i - 2) == '1')
                    dp[i] += dp[i - 2];
                if (s.charAt(i - 1) >= '0' && s.charAt(i - 1) <= '6' && s.charAt(i - 2) == '2')
                    dp[i] += dp[i - 2];
            }
            if (s.charAt(i - 1) == '0' && (i == 1 || (s.charAt(i - 2) != '1' && s.charAt(i - 2) != '2')))
                return 0;            
        }
        return dp[s.length()];
    }
}

File: Python/91. Decode Ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        ways = [0] * (len(s) + 1)
        ways[0] = 1
        for i in range(1, len(s) + 1):
            c = s[i - 1]
            if c >= '1' and c <= '9':
                ways[i] = ways[i - 1]
            if i > 1 and ((c >= '0' and c <= '9' and s[i - 2] == '1') or (c >= '0' and c <= '6' and s[i - 2] == '2')):
                ways[i] += ways[i - 2]
            if c == '0' and s[i - 2] != '1' and s[i - 2] != '2':
                return 0
        return 0 if len(s) == 0 else ways[len(s)]
